cmd/nsm-sriov-manager/nsm-sriov-manager.go

```diff
@@ -0,0 +1,418 @@
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/sirupsen/logrus"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	netDirectory    = "/sys/class/net/"
+	pciDirectory    = "/sys/bus/pci/devices"
+	sriovCapable    = "/sriov_totalvfs"
+	sriovConfigured = "/sriov_numvfs"
+	// nsmSRIOVDefaultNetworkServiceName defines a default network service name for all
+	// SRIOV VFs found on a host. Further editing might be required to map VF to real
+	// Network Services
+	nsmSRIOVDefaultNetworkServiceName = "networkservicemesh.io/sriov/nsm-default"
+)
+
+// VF describes a single instance of VF
+type VF struct {
+	NetworkService string `yaml:"networkService" json:"NetworkService"`
+	pciAddr        string // `yaml:"pciAddr" json:"pciAddr"`
+	ParentDevice   string `yaml:"parentDevice" json:"parentDevice"`
+	VFLocalID      int32  `yaml:"vfLocalID" json:"vfLocalID"`
+	pciVendor      string // `yaml:"pciVendor" json:"pciVendor"`
+	pciType        string // `yaml:"pciType" json:"pciType"`
+	iommuGroup     string // `yaml:"iommuGroup" json:"iommuGroup"`
+	VFIODevice     string `yaml:"vfioDevice" json:"vfioDevice"`
+}
+
+// VFs is map of ALL found VFs on a specific host kyed by PCI address
+type VFs struct {
+	vfs map[string]*VF
+}
+
+func newVFs() *VFs {
+	v := &VFs{}
+	vfs := map[string]*VF{}
+	v.vfs = vfs
+	return v
+}
+
+// Returns a list of SRIOV capable PF names as string
+func getSriovPfList() ([]string, error) {
+
+	sriovNetDevices := []string{}
+
+	netDevices, err := ioutil.ReadDir(netDirectory)
+	if err != nil {
+		logrus.Errorf("Error. Cannot read %s for network device names. Err: %v", netDirectory, err)
+		return sriovNetDevices, err
+	}
+
+	if len(netDevices) < 1 {
+		logrus.Errorf("Error. No network device found in %s directory", netDirectory)
+		return sriovNetDevices, err
+	}
+
+	for _, dev := range netDevices {
+		sriovFilePath := filepath.Join(netDirectory, dev.Name(), "device", "sriov_numvfs")
+		f, err := os.Lstat(sriovFilePath)
+		if err == nil {
+			if f.Mode().IsRegular() { // and its a regular file
+				sriovNetDevices = append(sriovNetDevices, dev.Name())
+			}
+		}
+	}
+
+	return sriovNetDevices, nil
+}
+
+func readLinkData(link string) (string, error) {
+	dirInfo, err := os.Lstat(link)
+	if err != nil {
+		return "", fmt.Errorf("Error. Could not get directory information %s with error: %v", link, err)
+	}
+
+	if (dirInfo.Mode() & os.ModeSymlink) == 0 {
+		return "", fmt.Errorf("Error. No symbolic link %s", link)
+	}
+
+	info, err := os.Readlink(link)
+	if err != nil {
+		return "", fmt.Errorf("Error. Cannot read symbolic %s with error: %+v", link, err)
+	}
+
+	return info, nil
+}
+
+//Reads DeviceName and gets PCI Addresses of VFs configured
+func discoverNetworks(discoveredVFs *VFs) error {
+
+	var pciVendor, pciType string
+	// Get a list of SRIOV capable NICs in the host
```

Comment:
This has implicit assumptions that:

- we only pass through SRIOV devices
- we pass through all SRIOV devices

(though of course you can edit the output to fix both cases).

Reply:
Right, that was the intention, to give a user a chance to customize config. I do not see other way to do it, unless to force a user to provide bunch of input parameters, but even with that, user still need to edit the config map for network service.